### PR TITLE
Open telemetry support

### PR DESCRIPTION
**Pull request by BrynCooke**
This PR adds instrumentation to emit the following spans:
* federation.request - total time to server a request.
* federation.validate - time to validate the federated query.
* federation.plan - time to plan the federated query.
* federation.execute - the total time executing query.
* federation.fetch - time fetching data from a subgraph.
* federation.postprocessing - time to render the response from all the subgraph data.

The instrumentation looks a little messy due to the requirement to set error status on spans on exception, but also if errors are added to the response.

### CODE REVIEW

**Review by trevor-scheer — CHANGES_REQUESTED** [review]
This is looking really great! I don't have any majorly blocking issues, and I think my couple larger suggestions are optional/food for thought anyway.

One separate comment I will make is that consistency in style would be nice. Things that come to mind are indentation, bracket spacing, semicolons, and newlines. Note that this is explicitly not enforced on this repo and it's certainly not correct everywhere, either, so do with that what you will. It's not something I'll go around and leaving nitpick comments about or block on.

File: gateway-js/package.json [hunk 1]
```diff
@@ -37,9 +37,13 @@
...
     "make-fetch-happen": "^8.0.0",
-    "pretty-format": "^26.0.0"
+    "pretty-format": "^26.0.0",
+    "@opentelemetry/api": "^0.20.0"
   },
   "peerDependencies": {
     "graphql": "^14.5.0 || ^15.0.0"
+  },
+  "devDependencies": {
+    "@opentelemetry/node": "^0.20.0"
```
[the remark this answers is not in the document]
Comment (BrynCooke) [reply gateway-js/package.json]: Done

File: gateway-js/src/executeQueryPlan.ts [hunk 2]
```diff
@@ -197,99 +224,113 @@ async function executeFetch<TContext>(
...
 
-  let entities: ResultMap[];
-  if (Array.isArray(results)) {
-    // Remove null or undefined entities from the list
-    entities = results.filter(isNotNullOrUndefined);
-  } else {
-    entities = [results];
-  }
+  return tracer.startActiveSpan(OpenTelemetrySpanNames.FETCH, async span => {
+    span.setAttribute("sevice", fetch.serviceName);
```
[the remark this answers is not in the document]
Comment (BrynCooke) [reply gateway-js/src/executeQueryPlan.ts]: Done

File: gateway-js/src/__tests__/gateway/opentelemetry.test.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        }
+      }
+    `;
+
+      try {
+        await execute(executor, source);
+      }
+      catch(err) {
+
+      }
```
[the remark this answers is not in the document]
Comment (BrynCooke) [reply gateway-js/src/__tests__/gateway/opentelemetry.test.ts]: Done

File: federation-integration-testsuite-js/src/snapshotSerializers/readableSpanArraySerializer.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            const parentSpanMirror = spanMap.get(
+                spans.find((s2) => s2.spanContext().spanId === s.parentSpanId)
+            );
+            if(Array.isArray(parentSpanMirror)) {
+                parentSpanMirror.push(spanMirror);
+            }
+            else {
+                parentSpanMirror.children.push(spanMirror);
+            }
+        });
```
[the remark this answers is not in the document]
Comment (BrynCooke) [reply federation-integration-testsuite-js/src/snapshotSerializers/readableSpanArraySerializer.ts]: I added some comments and also simplified the algorithm.
At least part of the issue was that I had named things badly. Hopefully it looks better now.

**Comment by trevor-scheer** [comment]
I'm less familiar with the nuances of lockfiles generated by npm@7, but I would expect a diff caused by the move of the `devDependencies`. This change would be represented after running an `npm install` now that they're in the correct place within the `package.json` root. Can you please give that a run and see if there's any change?

File: gateway-js/CHANGELOG.md [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 > The changes noted within this `vNEXT` section have not been released yet.  New PRs and commits which introduce changes should include an entry in this `vNEXT` section as part of their development.  When a release is being prepared, a new header will be (manually) created below and the appropriate changes within that release will be moved into the new section.
 
--  _Nothing yet! Stay tuned!_
+- OpenTelemetry support PR TBA
```
Comment (trevor-scheer) [inline gateway-js/CHANGELOG.md:7]: Let's update this with the PR # and link. I think the description is fine but if you feel there's anything worth adding here that's great too.

**Review by trevor-scheer — APPROVED** [review]
Aside from my `package-lock.json` concern, this looks ready to me!

File: gateway-js/src/index.ts [hunk 6]
```diff
@@ -928,142 +932,185 @@ export class ApolloGateway implements GraphQLService {
...
+        }
+        return errors || [];
+      }
+      catch (err) {
+        span.setStatus({ code:SpanStatusCode.ERROR });
+        throw err;
+      }
+      finally {
+        span.end();
+      }
```
[the remark this answers is not in the document]
Comment (BrynCooke) [reply gateway-js/src/index.ts:1109]: I was also wondering about such a helper function. It seems strange that they didn't bake this in to the OpenTelemetry API.
startActiveSpan actually takes a bunch more parameters, and I realised that I should have been using one of these overloads at least once.

Let's leave it for now, but agree it's not pretty.

[the remark this answers is not in the document]
Comment (trevor-scheer) [reply gateway-js/src/index.ts:1109]: Sgtm!

**Comment by BrynCooke** [comment]
Thanks for all the feedback @trevor-scheer !

**Pull request merged**